### PR TITLE
Allow XDG base directory to be used to store application data on Linux

### PR DESCRIPTION
**Pull request by memchr**
## Description
If `~/.renderdoc` does not exist, `~/.config/renderdoc` will be used as the application directory. If `~/.renderdoc` exists and is a directory, then it will be used instead.

This means that a new user will probably have `~/.config/renderdoc` as their application directory by default. There is currently no automated migration from ~/.renderdoc to `~/.config/renderdoc`. If users wish to migrate to the new path, they shall do so manually.

## Motivations
This change is to partially support the XDG Base Directory Spec to reduce the amount of top level hidden files in the HOME directory.

`~/.config` is a well-known and widely used analogue of `/etc` for user-specific configuration on major Linux distributions, and sometimes, albeit mistakenly, used on Mac OS. This path likely to remain unchanged for the next few decades, as it has for the last two. Many existing, widely deployed, projects use `~/.config` by default. For example, `gtk`, `mesa`, `qtcreator`, `chromium` project, `visual studio code`, `pip`, `pipx`, `clangd`, `CUPS`, etc. Therefore, it might be logically reasonable and common for the users to assume that renderdoc will also use this path.

The implementation is simple and minimal, the new preferred application path is hard-coded and does NOT depend on any new environment variables.

Albeit the spec states that `$XDG_CONFIG_HOME` should be used instead of hard-coding, this practice shall be strongly discouraged as it introduces additional complexity, maintenance cost, and potential for bugs by inviting variability in system configuration, especially on Linux where systems vary wildly in unpredictable ways. Thus, with these factors in mind, **Under NO circumstances I will consider with any request to use environment variables to support the XDG Base Directory Specification in this PR**

For simplicity, the cache, config or any logs would not be correspondingly moved to ~/.cache or ~/.local/state.

### CODE REVIEW

**Comment by baldurk** [comment]
I'm not sure I want to accept this. I've proposed this compromise before to people who wanted support for XDG and it was clearly rejected as an idea. In particular:

> Albeit the spec states that $XDG_CONFIG_HOME should be used instead of hard-coding, this practice shall be strongly discouraged as it introduces additional complexity, maintenance cost, and potential for bugs by inviting variability in system configuration, especially on Linux where systems vary wildly in unpredictable ways. Thus, with these factors in mind, This PR is NOT open to discussion for the use of environment variables to support XDG Base Directory Spec.

This is not my impression and quite the opposite - I've found that people are want the spec to be very strictly followed. I do agree with that recommendation but I've not seen that discouragement more widely stated.

So my concern is that this will be a compromise that no-one is happy with and so doesn't add sufficient value because it doesn't make the people wanting XDG happy, and those who don't care will have at best a neutral experience.

As a side note, as the author of a PR your position is not to say what discussion is and isn't allowed.

**Comment by memchr** [comment]
> As a side note, as the author of a PR your position is not to say what discussion is and isn't allowed.

I sincerely apologise for the language used here. My impression from #1741 is that the use of XDG environment variables is not welcome, hence the assertive and potentially offensive description to discourage other users from suggesting the use of environment variables. However, being uncivil is not my aim here. I will change the PR as you wish.

Again, my apologies for the impression and thank you for your time.

- https://github.com/baldurk/renderdoc/pull/1741#issuecomment-590957100
- https://github.com/baldurk/renderdoc/pull/1741#issuecomment-592477184

**Comment by memchr** [comment]
> This is not my impression and quite the opposite - I've found that people are want the spec to be very strictly followed. 

Has the objective changed? Is it currently acceptable for PRs to strictly implement specifications?

**Comment by baldurk** [comment]
Oh I see what you are getting at now. I'm sorry myself, I completely misunderstood where you were coming from, I thought you were saying that to me from yourself that you didn't want to discuss using the environment variables and not that you were mirroring my previous PR feedback. I still hold that same view but since I'm effectively the only reviewer on RenderDoc pull requests there's no need to give that kind of direction as no-one else will be doing any PR review.

Unfortunately as I say above I think those env vars are still the critical point in this. I'm mostly fine with this compromise which is why I originally suggested it on that PR but I don't think for people who want XDG support that this is any more acceptable than the current situation. If the compromise isn't acceptable to both sides then I'm not sure it's worthwhile to make it.

**Comment by memchr** [comment]
> those env vars are still the critical point in this

Is it acceptable to use the  `XDG_CONFIG_HOME` variable to allow the use of `$XDG_CONFIG_HOME/renderdoc`, but **if and only if** all of the following conditions are met?

- `~/.renderdoc` does not exist or is not writable
- `$XDG_CONFIG_HOME` environment variable is not empty
- `$XDG_CONFIG_HOME/renderdoc` exists and is writable

This means that users who want XDG support can explicitly set renderdoc to use XDG_CONFIG_HOME in a way that is mostly compliant and **leaves the majority of users unaffected**. This practice is often done when the project does not want to disrupt existing users.

As the current implementation may require a change to FileIO for all platforms to separate config, data and cache as specified in the spec, and I have not been able to test this on Mac or Windows, and for simplicity's sake, caches and data are stored indiscriminately in `$XDG_CONFIG_HOME/renderdoc` in this PR.

Not separating data and cache, which is not an uncommon practice, may be an acceptable compromise.

As in the past, changes will be kept to a minimum to avoid unreasonable increases in maintenance costs.

---

In short, mayhap this PR will improve the experience for users who prefer to use the XDG base directories, while having minimal impact on users who are indifferent to this kind matter.

**Comment by baldurk** [comment]
No I don't want to make any change that depends on an environment variable for any part of this behaviour.

**Comment by memchr** [comment]
> No I don't want to make any change that depends on an environment variable for any part of this behaviour.

I respect your decision, but may I clarify that this variable is not set by default on common Linux distributions, and even if it is, the other two conditions must also be met, **so code paths are highly unlikely to be executed unless the user explicitly allows it**. ?

**Comment by memchr** [comment]
If XDG_CONFIG_HOME is not to your liking, would a variable specific to renderdoc be acceptable?

**Comment by memchr** [comment]
> No I don't want to make any change that depends on an environment variable for any part of this behaviour

> people who want XDG support that this is any more acceptable than the current situation

> a compromise that no-one is happy


Given the above conclusions, is it reasonable to assume that we have reached an impasse on implementing XDG support in this project? May I ask what your suggestions are?

If this PR is to be closed, should there be a section in the contribution guidelines informing that XDG does not align with the project's objectives? This would clarify the reasoning behind not implementing XDG and redirect potentially wasted efforts to other areas.

Please excuse my intrusion and thanks again for your time and effort!

**Comment by baldurk** [comment]
You didn't give me a chance to reply to your earlier comment, but no a different environment variable wouldn't be any better.

It sounds like no there is no reasonable compromise to make. As above I would have been fine with the `.config` solution but it seems like even you don't want that and want to have the environment variable support. For that reason I will close this PR. I don't see any reason to put a specific mention of this in the documentation as it's not reasonable to enumerate every single minor change that isn't made.

**Comment by memchr** [comment]
>  you don't want that and want to have the environment variable support

My position regarding support for environment variables remains neutral, indicating that I do not possess any particular preference. 

> As above I would have been fine with the .config solution

I have maintained a local history for every change made, and if you prefer to consider the ".config" solution, I will revert back to the initial change. However, if you no longer wish to proceed with this pull request, please kindly refrain from replying.

Thank you.

**Comment by memchr** [comment]
> You didn't give me a chance to reply to your earlier comment, but no a different environment variable wouldn't be any better.

Sorry, as there is no reply to earlier messages, I jumped the gun.
